packages/miniflare/src/workers/d1/dumpSql.ts

```diff
@@ -0,0 +1,152 @@
+import { SqlStorage } from "@cloudflare/workers-types/experimental";
+
+export function* dumpSql(
+	db: SqlStorage,
+	options?: {
+		noSchema?: boolean;
+		noData?: boolean;
+		tables?: string[];
+	}
+) {
+	yield `PRAGMA defer_foreign_keys=TRUE;`;
+
+	// Empty set means include all tables
+	const filterTables = new Set(options?.tables || []);
+	const { noData, noSchema } = options || {};
+
+	// Taken from SQLite shell.c.in https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L8444-L8450
+	// @ts-ignore (SqlStorageStatement needs to be callable)
+	const tables_cursor = db.prepare(`
+    SELECT name, type, sql 
+      FROM sqlite_schema AS o 
+    WHERE (true) AND type=='table' 
+      AND sql NOT NULL 
+    ORDER BY tbl_name='sqlite_sequence', rowid;
+  `)();
+	const tables: any[] = Array.from(tables_cursor);
+
+	for (const { name: table, sql } of tables) {
+		if (filterTables.size > 0 && !filterTables.has(table)) continue;
+
+		if (table === "sqlite_sequence") {
+			if (!noSchema) yield `DELETE FROM sqlite_sequence;`;
+		} else if (table.startsWith(`sqlite_stat`)) {
+			// This feels like it should really appear _after_ the contents of sqlite_stat[1,4] but I'm choosing
+			// to match SQLite's dump output exactly so writing it immediately like they do.
+			if (!noSchema) yield `ANALYZE sqlite_schema;`;
+		} else if (sql.startsWith(`CREATE VIRTUAL TABLE`)) {
+			throw new Error(
+				`D1 Export error: cannot export databases with Virtual Tables (fts5)`
+			);
+		} else if (table.startsWith("_cf_") || table.startsWith("sqlite_")) {
+			continue;
+		} else {
+			// SQLite dump has an extremely weird behaviour where, if the table was explicitly
+			// quoted i.e. "Table", then in the dump it has `IF NOT EXISTS` injected. I don't understand
+			// why, but on the off chance there's a good reason to I am following suit.
+			if (sql.match(/CREATE TABLE ['"].*/)) {
+				if (!noSchema) yield `CREATE TABLE IF NOT EXISTS ${sql.substring(13)};`;
+			} else {
+				if (!noSchema) yield `${sql};`;
+			}
+		}
+
+		if (noData) continue;
+		const columns_cursor = db.exec(`PRAGMA table_info="${table}"`);
+		const columns = Array.from(columns_cursor);
+		const select = `SELECT ${columns.map((c) => c.name).join(", ")}
+                            FROM "${table}";`;
+		const rows_cursor = db.exec(select);
+		for (const dataRow of rows_cursor.raw()) {
+			const formattedCells = dataRow.map((cell: unknown, i: number) => {
+				const colType = columns[i].type;
+				const cellType = typeof cell;
+				if (cell === null) {
+					return "NULL";
+				} else if (colType === "INTEGER" || cellType === "number") {
+					return cell;
+				} else if (colType === "TEXT" || cellType === "string") {
+					// Reproduce `output_quoted_escaped_string` from https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L1799-L1862
+					let lfs = false;
+					let crs = false;
+					// First escape any ' as ''
+					let output_string = `'${(cell as string).replace(
```

Comment:
Nit: this could potentially be extracted to a separate function to make the `sqlDump` function more readable.

Reply:
Yeah I wrote this file while trying my best to copy line-by-line the functionality in shell.c.in, but I think it makes sense to then slightly restructure it for readability after the fact.